File: src/feats/ops/time_parser.rs
//! 时间解析算子：将多种时间格式解析为稳定整数特征。
use super::{CustomOp, Fv};
use chrono::{DateTime, Datelike, NaiveDate, NaiveDateTime, Timelike, Utc};

/// 输入时间格式。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum InputFormat {
    Auto,
    EpochS,
    EpochMs,
    Rfc3339,
    Strftime,
}

impl InputFormat {
    fn parse(raw: &str) -> Result<Self, String> {
        match raw {
            "auto" => Ok(Self::Auto),
            "epoch_s" => Ok(Self::EpochS),
            "epoch_ms" => Ok(Self::EpochMs),
            "rfc3339" => Ok(Self::Rfc3339),
            "strftime" => Ok(Self::Strftime),
            other => Err(format!("TimeParser: unsupported input_format '{}'", other)),
        }
    }
}

/// 输出字段。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum OutputField {
    TimestampS,
    Year,
    Month,
    Day,
    Hour,
    Minute,
    Weekday,
    DayOfYear,
    Yyyymmdd,
    MinuteOfDay,
}

impl OutputField {
    fn parse(raw: &str) -> Result<Self, String> {
        match raw {
            "timestamp_s" => Ok(Self::TimestampS),
            "year" => Ok(Self::Year),
            "month" => Ok(Self::Month),
            "day" => Ok(Self::Day),
            "hour" => Ok(Self::Hour),
            "minute" => Ok(Self::Minute),
            "weekday" => Ok(Self::Weekday),
            "day_of_year" => Ok(Self::DayOfYear),
            "yyyymmdd" => Ok(Self::Yyyymmdd),
            "minute_of_day" => Ok(Self::MinuteOfDay),
            other => Err(format!("TimeParser: unsupported output '{}'", other)),
        }
    }
}

/// 将时间字符串、秒/毫秒时间戳解析为整数时间特征。
pub struct TimeParser {
    input_format: InputFormat,
    output: OutputField,
    formats: Vec<String>,
    default_val: i32,
}

impl TimeParser {
    /// 创建时间解析算子。
    pub fn new(
        input_format: &str,
        output: &str,
        formats: Vec<String>,
        default_val: i32,
    ) -> Result<Self, String> {
        Ok(Self {
            input_format: InputFormat::parse(input_format)?,
            output: OutputField::parse(output)?,
            formats,
            default_val,
        })
    }

    fn parse_input(&self, input: &Fv) -> Option<DateTime<Utc>> {
        match self.input_format {
            InputFormat::EpochS => numeric_i64(input).and_then(datetime_from_epoch_s),
            InputFormat::EpochMs => numeric_i64(input).and_then(datetime_from_epoch_ms),
            InputFormat::Rfc3339 => string_value(input).and_then(parse_rfc3339),
            InputFormat::Strftime => {
                string_value(input).and_then(|s| parse_with_formats(s, &self.formats))
            }
            InputFormat::Auto => self.parse_auto(input),
        }
    }

    fn parse_auto(&self, input: &Fv) -> Option<DateTime<Utc>> {
        if let Some(value) = numeric_i64(input) {
            return parse_auto_epoch(value);
        }
        let value = string_value(input)?;
        if let Ok(number) = value.parse::<i64>() {
            return parse_auto_epoch(number);
        }
        parse_rfc3339(value)
            .or_else(|| parse_with_formats(value, &self.formats))
            .or_else(|| parse_common_formats(value))
    }

    fn project(&self, dt: DateTime<Utc>) -> Option<i32> {
        match self.output {
            OutputField::TimestampS => i32::try_from(dt.timestamp()).ok(),
            OutputField::Year => Some(dt.year()),
            OutputField::Month => Some(dt.month() as i32),
            OutputField::Day => Some(dt.day() as i32),
            OutputField::Hour => Some(dt.hour() as i32),
            OutputField::Minute => Some(dt.minute() as i32),
            OutputField::Weekday => Some(dt.weekday().num_days_from_monday() as i32),
            OutputField::DayOfYear => Some(dt.ordinal() as i32),
            OutputField::Yyyymmdd => {
                Some(dt.year() * 10_000 + dt.month() as i32 * 100 + dt.day() as i32)
            }
            OutputField::MinuteOfDay => Some((dt.hour() * 60 + dt.minute()) as i32),
        }
    }
}

impl CustomOp for TimeParser {
    fn name(&self) -> &str {
        "TimeParser"
    }

    fn process(&self, inputs: &[Fv]) -> Result<Fv, String> {
        let value = inputs
            .first()
            .and_then(|input| self.parse_input(input))
            .and_then(|dt| self.project(dt))
            .unwrap_or(self.default_val);
        Ok(Fv::Int(value))
    }

    fn process_batch(&self, inputs: &[&[Fv]], n_rows: usize) -> Result<Vec<Fv>, String> {
        let col = inputs[0];
        let mut results = Vec::with_capacity(n_rows);
        for input in col.iter().take(n_rows) {
            let value = self
                .parse_input(input)
                .and_then(|dt| self.project(dt))
                .unwrap_or(self.default_val);
            results.push(Fv::Int(value));
        }
        Ok(results)
    }
}

/// 从 YAML params 创建 TimeParser 算子。
pub fn create(params: &serde_yaml::Value) -> Result<Box<dyn CustomOp>, String> {
    let input_format = params
        .get("input_format")
        .and_then(|v| v.as_str())
        .unwrap_or("auto");
    let output = params
        .get("output")
        .and_then(|v| v.as_str())
        .unwrap_or("timestamp_s");
    let formats = params
        .get("formats")
        .and_then(|v| v.as_sequence())
        .map(|seq| {
            seq.iter()
                .filter_map(|v| v.as_str().map(str::to_string))
                .collect()
        })
        .unwrap_or_default();
    let default_val = params
        .get("default_val")
        .and_then(|v| v.as_i64())
        .unwrap_or(0) as i32;
    Ok(Box::new(TimeParser::new(
        input_format,
        output,
        formats,
        default_val,
    )?))
}

fn numeric_i64(input: &Fv) -> Option<i64> {
    match input {
        Fv::Int(value) => Some(i64::from(*value)),
        Fv::Float(value) if value.is_finite() => Some(*value as i64),
        Fv::Str(value) => value.parse::<i64>().ok(),
        _ => None,
    }
}

fn string_value(input: &Fv) -> Option<&str> {
    match input {
        Fv::Str(value) => {
            let trimmed = value.trim();
            (!trimmed.is_empty()).then_some(trimmed)
        }
        _ => None,
    }
}

fn parse_auto_epoch(value: i64) -> Option<DateTime<Utc>> {
    if value.abs() >= 100_000_000_000 {
        datetime_from_epoch_ms(value)
    } else {
        datetime_from_epoch_s(value)
    }
}

fn datetime_from_epoch_s(value: i64) -> Option<DateTime<Utc>> {
    DateTime::from_timestamp(value, 0)
}

fn datetime_from_epoch_ms(value: i64) -> Option<DateTime<Utc>> {
    DateTime::from_timestamp_millis(value)
}

fn parse_rfc3339(value: &str) -> Option<DateTime<Utc>> {
    DateTime::parse_from_rfc3339(value)
        .ok()
        .map(|dt| dt.with_timezone(&Utc))
}

fn parse_with_formats(value: &str, formats: &[String]) -> Option<DateTime<Utc>> {
    for fmt in formats {
        if let Some(dt) = parse_with_format(value, fmt) {
            return Some(dt);
        }
    }
    None
}

fn parse_common_formats(value: &str) -> Option<DateTime<Utc>> {
    for fmt in COMMON_FORMATS {
        if let Some(dt) = parse_with_format(value, fmt) {
            return Some(dt);
        }
    }
    None
}

fn parse_with_format(value: &str, fmt: &str) -> Option<DateTime<Utc>> {
    if let Ok(dt) = DateTime::parse_from_str(value, fmt) {
        return Some(dt.with_timezone(&Utc));
    }
    if let Ok(dt) = NaiveDateTime::parse_from_str(value, fmt) {
        return Some(dt.and_utc());
    }
    if let Ok(date) = NaiveDate::parse_from_str(value, fmt) {
        return date.and_hms_opt(0, 0, 0).map(|dt| dt.and_utc());
    }
    None
}

const COMMON_FORMATS: &[&str] = &[
    "%Y-%m-%d %H:%M:%S",
    "%Y-%m-%d %H:%M",
    "%Y-%m-%d",
    "%Y/%m/%d %H:%M:%S",
    "%Y/%m/%d %H:%M",
    "%Y/%m/%d",
    "%Y%m%d%H%M%S",
    "%Y%m%d",
];

#[cfg(test)]
mod tests {
    use super::*;
    use crate::feats::ops::CustomOp;

    #[test]
    fn parses_rfc3339_to_hour_utc() {
        let op = TimeParser::new("rfc3339", "hour", vec![], -1).unwrap();
        let result = op.process(&[Fv::Str("2026-06-29T10:30:00+08:00".into())]);
        assert_eq!(result.unwrap(), Fv::Int(2));
    }

    #[test]
    fn parses_custom_format_to_weekday() {
        let op = TimeParser::new("strftime", "weekday", vec!["%Y.%m.%d %H:%M".into()], -1).unwrap();
        assert_eq!(
            op.process(&[Fv::Str("2026.06.29 12:00".into())]).unwrap(),
            Fv::Int(0)
        );
    }

    #[test]
    fn parses_epoch_ms_input() {
        let op = TimeParser::new("epoch_ms", "yyyymmdd", vec![], -1).unwrap();
        assert_eq!(
            op.process(&[Fv::Str("1782727200000".into())]).unwrap(),
            Fv::Int(20260629)
        );
    }

    #[test]
    fn falls_back_to_default_on_invalid_input() {
        let op = TimeParser::new("auto", "day_of_year", vec![], 366).unwrap();
        assert_eq!(op.process(&[Fv::Str("bad".into())]).unwrap(), Fv::Int(366));
    }
}

File: python/src/train/ops/time_parser.py
from __future__ import annotations

"""时间解析算子：将多种时间格式解析为稳定整数特征。"""

from datetime import UTC, datetime
from typing import Any

from . import register_op

COMMON_FORMATS = (
    "%Y-%m-%d %H:%M:%S",
    "%Y-%m-%d %H:%M",
    "%Y-%m-%d",
    "%Y/%m/%d %H:%M:%S",
    "%Y/%m/%d %H:%M",
    "%Y/%m/%d",
    "%Y%m%d%H%M%S",
    "%Y%m%d",
)


@register_op("TimeParser")
class TimeParser:
    """Parse string or epoch time values into integer time features."""

    def __init__(
        self,
        input_format: str = "auto",
        output: str = "timestamp_s",
        formats: list[str] | None = None,
        default_val: int = 0,
    ) -> None:
        self.input_format = input_format
        self.output = output
        self.formats = formats or []
        self.default_val = default_val
        self._validate()

    @classmethod
    def from_config(cls, params: dict) -> TimeParser:
        return cls(
            input_format=str(params.get("input_format", "auto")),
            output=str(params.get("output", "timestamp_s")),
            formats=[str(fmt) for fmt in params.get("formats", [])],
            default_val=int(params.get("default_val", 0)),
        )

    def _validate(self) -> None:
        if self.input_format not in {"auto", "epoch_s", "epoch_ms", "rfc3339", "strftime"}:
            raise ValueError(f"TimeParser: unsupported input_format '{self.input_format}'")
        if self.output not in {
            "timestamp_s",
            "year",
            "month",
            "day",
            "hour",
            "minute",
            "weekday",
            "day_of_year",
            "yyyymmdd",
            "minute_of_day",
        }:
            raise ValueError(f"TimeParser: unsupported output '{self.output}'")

    def process(self, inputs: list[Any]) -> int:
        dt = self._parse_input(inputs[0] if inputs else None)
        if dt is None:
            return self.default_val
        return self._project(dt)

    def process_batch(self, inputs: list[Any]) -> list[int]:
        vals = inputs[0]
        return [self.process([val]) for val in vals]

    def _parse_input(self, value: Any) -> datetime | None:
        if self.input_format == "epoch_s":
            return _datetime_from_epoch(value, scale=1)
        if self.input_format == "epoch_ms":
            return _datetime_from_epoch(value, scale=1000)
        if self.input_format == "rfc3339":
            return _parse_rfc3339(value)
        if self.input_format == "strftime":
            return _parse_with_formats(value, self.formats)
        return self._parse_auto(value)

    def _parse_auto(self, value: Any) -> datetime | None:
        number = _to_int(value)
        if number is not None:
            scale = 1000 if abs(number) >= 100_000_000_000 else 1
            return _datetime_from_epoch(number, scale=scale)
        return (
            _parse_rfc3339(value)
            or _parse_with_formats(value, self.formats)
            or _parse_with_formats(value, COMMON_FORMATS)
        )

    def _project(self, dt: datetime) -> int:
        if self.output == "timestamp_s":
            value = int(dt.timestamp())
            if value < -(2**31) or value > 2**31 - 1:
                return self.default_val
            return value
        if self.output == "year":
            return dt.year
        if self.output == "month":
            return dt.month
        if self.output == "day":
            return dt.day
        if self.output == "hour":
            return dt.hour
        if self.output == "minute":
            return dt.minute
        if self.output == "weekday":
            return dt.weekday()
        if self.output == "day_of_year":
            return int(dt.strftime("%j"))
        if self.output == "yyyymmdd":
            return dt.year * 10_000 + dt.month * 100 + dt.day
        if self.output == "minute_of_day":
            return dt.hour * 60 + dt.minute
        return self.default_val


def _to_int(value: Any) -> int | None:
    if isinstance(value, bool) or value is None:
        return None
    if isinstance(value, int):
        return value
    if isinstance(value, float) and value.is_integer():
        return int(value)
    if isinstance(value, str):
        text = value.strip()
        if not text:
            return None
        try:
            return int(text)
        except ValueError:
            return None
    return None


def _datetime_from_epoch(value: Any, scale: int) -> datetime | None:
    number = _to_int(value)
    if number is None:
        return None
    try:
        return datetime.fromtimestamp(number / scale, tz=UTC)
    except (OSError, OverflowError, ValueError):
        return None


def _parse_rfc3339(value: Any) -> datetime | None:
    if not isinstance(value, str):
        return None
    text = value.strip()
    if not text:
        return None
    if text.endswith("Z"):
        text = f"{text[:-1]}+00:00"
    try:
        parsed = datetime.fromisoformat(text)
    except ValueError:
        return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=UTC)
    return parsed.astimezone(UTC)


def _parse_with_formats(value: Any, formats: list[str] | tuple[str, ...]) -> datetime | None:
    if not isinstance(value, str):
        return None
    text = value.strip()
    if not text:
        return None
    for fmt in formats:
        try:
            parsed = datetime.strptime(text, fmt)  # noqa: DTZ007
        except ValueError:
            continue
        if parsed.tzinfo is None:
            parsed = parsed.replace(tzinfo=UTC)
        return parsed.astimezone(UTC)
    return None
